NSrtm.Core/Pgm/GridGraph/PgmGridGraphInFile.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using JetBrains.Annotations;
+
+namespace NSrtm.Core.Pgm.GridGraph
+{
+    public sealed class PgmGridGraphInFile : IPgmGridGraph, IDisposable
+    {
+        private readonly FileStream _fileStream;
+        private readonly Object _thisLock = new Object();
+        private readonly PgmDataDescription _pgmParameters;
+
+        public PgmGridGraphInFile([NotNull] FileStream stream, PgmDataDescription pgmParameters)
+        {
+            if (stream == null) throw new ArgumentNullException("stream");
+            _fileStream = stream;
+            _pgmParameters = pgmParameters;
+        }
+
+        private double getUndulationFrom(int position)
+        {
+            lock (_thisLock)
+            {
+                _fileStream.Seek(position, SeekOrigin.Begin);
+                UInt16 rawData = (UInt16)(_fileStream.ReadByte() << 8 | _fileStream.ReadByte());
+                return rawData.ToEgmFormat(Parameters);
+            }
+        }
+
+        [SuppressMessage("Microsoft.Usage", "CA2208:InstantiateArgumentExceptionsCorrectly",
+            Justification = "Simple range checking - it makes no sense to create a separate method")]
+        private int getClosestPosition(double latitude, double longitude)
+        {
+            int latPoints = (int)Math.Round((_pgmParameters.OriginLat - latitude) * _pgmParameters.LatitudeIncrement);
```

Comment:
And please where code/logic is reusable use static methods  or composition to share it (Ctrl+C is feature of last resort for 20 expresions stuck together in 7 lines of code being: a) identical to second occurence, b) impossible to change spearately).